src/pools/cluster/dynamic.ts

```diff
@@ -61,7 +61,8 @@ export class DynamicClusterPool<
     // All workers are busy, create a new worker
     const worker = this.createAndSetupWorker()
     this.registerWorkerMessageListener<Data>(worker, message => {
-      if (message.kill) {
+      const tasksInProgress = this.tasks.get(worker)
+      if (message.kill && !tasksInProgress) {
```

Comment:
Hum, you forbid to destroy a worker with a task in progress ... that's not good, you shall have to try to: 
+ terminate or interrupt the current task;
+ handle task termination error;
+ conditionally terminate the worker.

I would prefer to follow POSIX kill semantic:
+ kill -> SIGINT
+ killHard -> SIGKILL: terminate the task and destroy the worker unconditionally

Reply:
Using other semantics here could be a breaking change, could it?
Maybe we should re-type `MessageValue`'s `kill` to a string literal :thinking: 
`kill?: 'SIGINT' | 'SIGKILL'`

Reply:
I am not sure I got your point @jerome-benoit , the desired behavior is to avoid worker/process termination if a task is in progress

Reply:
> Using other semantics here could be a breaking change, could it?
> Maybe we should re-type `MessageValue`'s `kill` to a string literal 🤔
> `kill?: 'SIGINT' | 'SIGKILL'`

That PR already change the semantic. 

On the implementation side, I would prefer to have 
+ an ```
enum KillSignal {
	SIGINT: 1
    SIGKILL: 2
}
```
+ pass the signal to the message { kill: KillSignal.SIGNINT } and do the matching code to fulfil the request.

Reply:
Please be patient with enums! Always use `const enum` and prefer to use enums only if you use them internally or have a really good reason.
enums were introduced to typescript the old days but are not really well designed in ts
string literals solve the same intention but have a better performance and don't produce boilerplate code for the lib folder.

Reply:
> I am not sure I got your point @jerome-benoit , the desired behavior is to avoid worker/process termination if a task is in progress

The desired behaviour for one specific need ;) 

I do need to unconditionally kill worker and its tasks and with that PR, I now can't by using the existing message passing API.

Reply:
Ok now I got it.
@jerome-benoit  from the worker code when would you would send a SIGINT and when you would send a SIGKILL?

I would prefer anyway to handle on a discussion thread and to implement on another PR

Reply:
> Ok now I got it.
> @jerome-benoit from the worker code when would you would send a SIGINT and when you would send a SIGKILL?
> 
> I would prefer anyway to handle on a discussion thread and to implement on another PR

As long as the breaking change here reinstantiate the old behaviour is a not compatible manner, that's fine. 

We just need to talk a bit of the kill semantic first, there's three major use case: 
+ kill if worker not in use: SIGUSE
+ kill task -> worker unconditionally : SIGINT
(+ kill gracefully task -> worker: SIGTERM, ask for task to terminate, empty the task queue, etc. with error handling at each step)

Reply:
So @jerome-benoit  we could add another option to the worker constructor ( something like killSignal ) and then decide to kill or not the worker based on that, and as default we could kill unconditionally so we are also fully retro compatible !!!
@Shinigami92  hopefully the parenthesis into this comment are ok :P

Reply:
I don't care about your parenthesis and typos in issues and comments :wink:  
I just keep the git source files clean and organized, that's what important is

Reply:
:P you are right, just kidding @Shinigami92 <3

Reply:
> So @jerome-benoit we could add another option to the worker constructor ( something like killSignal ) and then decide to kill or not the worker based on that, and as default we could kill unconditionally so we are also fully retro compatible !!!

Yep, a kill default signal or kill default behaviour tunable. 
I do not care about the retro compatibility, the current default behaviour have issues and should be changed to cope with the most common use cases. 
I just want to be able the change the default behaviour if needed for the sake of backward compatibility and less common use cases. Just be able to come back to the previous default behaviour.

Reply:
I'm with @jerome-benoit 
Better change it now, instead of stacking versions
A major release is a major release and a breaking change is totally okay for a major release
That's for what versioning is for :shrug:

Reply:
That's fine, I will implement it into this branch as soon as I can